org.adempiere.ui.zk/WEB-INF/src/org/adempiere/webui/adwindow/ToolbarCustomButton.java

```diff
@@ -104,7 +107,77 @@ public void dynamicDisplay(boolean forceValidation) {
 
 		toolbarButton.setVisible(visible);
 	}
-	
+
+	public void pressedLogic() {
+		if (toolbarButton.getParent() == null)
+			return;
+
+		String pressedLogic = mToolbarButton.getPressedLogic();
+
+		if (Util.isEmpty(pressedLogic, true))
+			return;
+
+		ADWindow window = ADWindow.get(windowNo);
+
+		if (window == null)
+			return;
+
+		IADTabpanel adTabpanel = window.getADWindowContent().getADTab().getSelectedTabpanel();
+
+		if (adTabpanel == null || adTabpanel.getRecord_ID() <= 0)
+			return;
+
+		boolean isPressed = validateLogic(pressedLogic, adTabpanel.getTabNo());
+		((ToolBarButton) toolbarButton).setPressed(isPressed);
+	}
+
+	public void readOnlyLogic() {
+		if (toolbarButton.getParent() == null)
+			return;
+
+		String readOnlyLogic = mToolbarButton.getReadOnlyLogic();
+
+		if (Util.isEmpty(readOnlyLogic, true))
+			return;
+
+		ADWindow window = ADWindow.get(windowNo);
+
+		if (window == null)
+			return;
+
+		IADTabpanel adTabpanel = window.getADWindowContent().getADTab().getSelectedTabpanel();
+
+		if (adTabpanel == null || adTabpanel.getRecord_ID() <= 0)
+			return;
+
+		boolean isReadOnly = validateLogic(readOnlyLogic, adTabpanel.getTabNo());
+
+		toolbarButton.setDisabled(isReadOnly);
+	}
+
+	private boolean validateLogic(String logic, int tabNo) {
+		boolean isSQL = false;
+		boolean isValid = false;
+
+		if (logic.startsWith("@SQL=")) {
+			logic = logic.substring(5, logic.length());
+			isSQL = true;
+		}
+
+		logic = Env.parseContext(Env.getCtx(), windowNo, tabNo, logic, false);
+
+		if (isSQL) {
+			int result = DB.getSQLValue(null, logic);
+
+			if (result == 1)
+				isValid = true;
```

Comment:
This is inconsistent with the use of SQL logic in other part of the application. Preferably, should use Evaluator.parseSQLLogic instead.

Reply:
Changed to use Evaluator.parseSQLLogic